docs/proposals/koordlet/20221018-koordlet-support-cgroups-v2.md

```diff
@@ -0,0 +1,324 @@
+  * [Proposal](#proposal)
+    * [User Stories](#user-stories)
+      * [Story 1](#story-1)
+      * [Story 2](#story-2)
+    * [Design Principles](#design-principles)
+    * [API](#api)
+      * [Cgroup File](#cgroup-file)
+      * [Resource Executor](#resource-executor)
+      * [Update](#update)
+        * [Batch &amp; Cacheable Update](#batch--cacheable-update)
+      * [Read](#read)
+  * [Alternatives](#alternatives)
+  * [Implementation History](#implementation-history)
+
+<!--te-->
+
+## Glossary
+
+Control Group v2 (cgroup v2): https://docs.kernel.org/admin-guide/cgroup-v2.html
+
+## Summary
+
+To support cgroups-v2 in koordlet, a module for operating system files is proposed for collecting resource utilization and applying resource management strategies on various system environments.
+
+## Motivation
+
+The Linux cgroups-v2 mechanism provided by the 5.x kernels tends to mature, provides memcg QoS, PSI, iocost, and other capabilities in container resource isolation and monitoring, and unifies the mapping of containers to different cgroup subsystems. Kubernetes has enabled cgroups-v2 by default since v1.25, providing Kubelet MemoryQoS and other functional support.
+In the Koordinator <= v0.7, koordlet implements resource utilization collection, container resource management and dynamic QoS strategies based on Linux cgroups-v1 common subsystems such as cpu, cpuacct, cpuset and memcg. To exploit the kernel's high-version features and get compatible with Kubelet (>=v1.25), the koordlet should be able to run on the cgroups-v2 environment. On the other hand, the koordlet should keep the resource collection and management functionalities between cgroups-v1 and cgroups-v2 consistent.
+Therefore, the module Resource Executor is provided to cover up the operations on system files on different environments. Other modules can simply call the resource executor when they want to read or write any cgroup resource, no longer perceiving the underlying differences of various systems (e.g. cgroups-v1 vs. cgroups-v2).
+
+### Goals
+
+- Define the resource executor APIs.
+- Give examples about how to operate cgroups-v1 and v2.
+
+### Non-Goals/Future Work
+
+- Design new resource management strategies based on cgroups-v2.
+
+## Proposal
+
+### User Stories
+
+#### Story 1
+
+As a cluster administrator, I want to deploy the koordinator across different versions of kubernetes which may use cgroups-v2 by default.
+
+#### Story 2
+
+As a koordinator developer, I want to run the koordlet on the linux kernel 5.x that enables cgroups-v2.
+
+### Design Principles
+
+How we expect to operate cgroups:
+
+1. The cgroup path can be derived from the pod meta.
+2. To reduce IO cost, cacheable update since many updates take the same values.
+3. Uniformly operate both the cgroups-v1 files and the similar cgroups-v2 files.
+4. Extensible to operate VM-based container cgroups in guest OS.
+
+### API
+
+![image](/docs/images/resource-executor.svg)
```

Comment:
1. make cacheExecutor as global
2. cacheExecutor should be a util or a independent go rountine
3. use one `new method` to include v1/v2 generator
4. cgroup resource updator, default resource updator
5. all write operation should be done by cacheExecutor
6. reader is an independent util

Reply:
Fixed